File: apps/rux/frontend/src/viewport/binaryPoints.ts
// SPDX-FileCopyrightText: 2026 Povl Filip Sonne-Frederiksen
//
// SPDX-License-Identifier: GPL-3.0-or-later

/**
 * RUXP v1 — the binary point transport.
 *
 * `docs/gui/binary-points.md` is normative; this module is a reader for exactly
 * that format and nothing else. The point of RUXP is that a page of points
 * reaches `THREE.BufferAttribute` without being touched per point in
 * JavaScript, so {@link RuxpPage.view} hands back a typed array constructed
 * *directly over the received `ArrayBuffer`* — never a copy, never a loop.
 *
 * ## Little-endian only
 *
 * The header is read through a `DataView` with an explicit `littleEndian: true`,
 * so it decodes correctly anywhere. **The payload does not.** A typed-array view
 * uses the host's byte order and offers no way to override it, so on a
 * big-endian host every `f32` coordinate and every `u32` label in a page would
 * decode as silent garbage — no throw, no clue, just a scrambled cloud.
 *
 * That is the format's deliberate non-goal, not an oversight here
 * (`docs/gui/binary-points.md` § Endianness): every platform `rux gui` and its
 * browser client run on is little-endian. If a big-endian client ever matters,
 * the fix is a `flags` bit advertising byte order plus a byte-swapping slow
 * path — and this parser already refuses any page with an unknown flag set, so
 * such a page cannot be misread by today's code.
 *
 * ## Everything malformed is a throw
 *
 * A truncated, over-long, misaligned or unknown-version page is rejected with a
 * message naming the number that disagreed. The alternative — constructing a
 * view anyway — is either a bare `RangeError` from the typed-array constructor
 * or, worse, a view onto the wrong bytes. A caller that wants to survive a bad
 * page catches this and falls back to `format=json`.
 */

/** Value type of one field's section. The three codes v1 defines. */
export type RuxpFieldType = 'f32' | 'u8' | 'u32';

/** A typed-array view over one field's section, in whatever type it declares. */
export type RuxpView = Float32Array | Uint8Array | Uint32Array;

/** One entry of the header's field table. */
export interface RuxpField {
  /** ASCII, NUL-padding stripped — `xyz`, `rgb`, `normal`, `label`. */
  name: string;
  type: RuxpFieldType;
  /** Values per point: 3 for `xyz`, 1 for `label`. */
  components: number;
  /** Absolute offset of this field's section from the start of the buffer. */
  byteOffset: number;
}

/** A parsed page. Views are lazy but cached, so repeated lookups are free. */
export interface RuxpPage {
  /** Format version. Always 1 — anything else was refused during parsing. */
  version: number;
  /** Points in *this* page. */
  count: number;
  /** Index of this page's first point within the cloud. */
  offset: number;
  /** Points in the whole cloud. */
  total: number;
  /** Field table, in header order. */
  fields: RuxpField[];
  /**
   * Zero-copy view onto a field's section, or `null` when the page has no such
   * field. The same view object is returned on every call for a given name.
   */
  view(name: string): RuxpView | null;
}

/** `"RUXP"` — the four bytes every page starts with. */
const MAGIC = [0x52, 0x55, 0x58, 0x50] as const;

/** Header bytes before the field table. `header_size` is this + 16 per field. */
const FIXED_HEADER_BYTES = 40;
const FIELD_DESCRIPTOR_BYTES = 16;

/** The only version this reader understands. A v2 page must be refused. */
const SUPPORTED_VERSION = 1;

/**
 * Flag bits this reader knows how to honour.
 *
 * v1 defines none, so any bit set is a page written by something that expects
 * the reader to do something it has never heard of — byte-swapping, say, or
 * decoding quantised positions. Refusing is the whole reason the field exists.
 */
const KNOWN_FLAGS = 0;

const TYPE_SIZES: Record<RuxpFieldType, number> = { f32: 4, u8: 1, u32: 4 };

/**
 * Cheap, allocation-free "is this a RUXP page at all?".
 *
 * Safe on a short or empty buffer, because the interesting caller is the one
 * asking whether a `format=binary` response body is really binary — an old
 * server, or a proxy, can answer 200 with something else entirely.
 */
export function isRuxp(buffer: ArrayBuffer): boolean {
  if (buffer.byteLength < MAGIC.length) return false;
  const bytes = new Uint8Array(buffer, 0, MAGIC.length);
  return MAGIC.every((byte, index) => bytes[index] === byte);
}

/**
 * Parse a RUXP page, or throw explaining which rule it broke.
 *
 * Nothing is read out of the payload here: the whole body is validated against
 * the header first, and only then are views handed out. That ordering is what
 * makes {@link RuxpPage.view} unconditionally safe for a caller to use without
 * re-checking bounds.
 */
export function parseRuxp(buffer: ArrayBuffer): RuxpPage {
  if (!isRuxp(buffer)) {
    throw new Error(
      `RUXP: bad magic — expected "RUXP", buffer is ${describeStart(buffer)}`,
    );
  }
  if (buffer.byteLength < FIXED_HEADER_BYTES) {
    throw new Error(
      `RUXP: buffer is ${buffer.byteLength} bytes, too short for the ` +
        `${FIXED_HEADER_BYTES}-byte fixed header`,
    );
  }

  const header = new DataView(buffer);

  const version = header.getUint16(4, true);
  if (version !== SUPPORTED_VERSION) {
    // Refused, not guessed at: a v2 exists precisely because it changed
    // something this code would otherwise misread (`docs/gui/binary-points.md`
    // § "Not in v1").
    throw new Error(`RUXP: unsupported version ${version}, this client reads v${SUPPORTED_VERSION}`);
  }

  const headerSize = header.getUint16(6, true);
  const flags = header.getUint32(8, true);
  const fieldCount = header.getUint32(12, true);
  const count = header.getUint32(16, true);
  const offset = readU64(header, 24, 'offset');
  const total = readU64(header, 32, 'total');

  if (flags !== KNOWN_FLAGS) {
    throw new Error(`RUXP: unknown flags 0x${flags.toString(16)} set; v1 defines none`);
  }

  const expectedHeaderSize = FIXED_HEADER_BYTES + FIELD_DESCRIPTOR_BYTES * fieldCount;
  if (headerSize !== expectedHeaderSize) {
    throw new Error(
      `RUXP: header_size ${headerSize} disagrees with field_count ${fieldCount} ` +
        `(expected ${expectedHeaderSize})`,
    );
  }
  if (headerSize > buffer.byteLength) {
    throw new Error(
      `RUXP: header_size ${headerSize} exceeds the ${buffer.byteLength}-byte buffer`,
    );
  }

  // Pass 1: read the field table and work out how long the body must be.
  //
  // Split from the bounds checking below so a truncated response is diagnosed
  // as a truncated response. Doing both in one loop reports whichever field
  // happens to sit last as "outside the payload", which sends the reader
  // hunting for a layout bug that is not there.
  const fields: RuxpField[] = [];
  const seen = new Set<string>();
  let payloadBytes = 0;

  for (let index = 0; index < fieldCount; index += 1) {
    const base = FIXED_HEADER_BYTES + index * FIELD_DESCRIPTOR_BYTES;
    const name = readName(buffer, base);
    const type = typeFromCode(header.getUint8(base + 8), name);
    const components = header.getUint8(base + 9);
    const byteOffset = header.getUint32(base + 12, true);

    if (components === 0) {
      // A zero-component field contributes no bytes and no meaning, so it
      // cannot be what the server intended to send — and it would hand the
      // caller an empty view for an attribute it believes it received.
      throw new Error(`RUXP: field '${name}' declares 0 components`);
    }
    if (seen.has(name)) {
      throw new Error(`RUXP: duplicate field '${name}' — view() would be ambiguous`);
    }
    seen.add(name);

    payloadBytes += count * components * TYPE_SIZES[type];
    fields.push({ name, type, components, byteOffset });
  }

  // Both directions are protocol errors. Short would mean reading off the end;
  // long means the sender and this reader disagree about the layout, and the
  // bytes that agree cannot be told from the bytes that do not.
  const bodyLength = headerSize + payloadBytes;
  if (bodyLength !== buffer.byteLength) {
    throw new Error(
      `RUXP: body_length ${bodyLength} (header ${headerSize} + payload ${payloadBytes}) ` +
        `does not match the received ${buffer.byteLength} bytes`,
    );
  }

  // Pass 2: every section must lie inside the payload and be legally aligned.
  // The body is known to be the right *size* by here, so anything caught below
  // is a header that misplaces a section within it.
  for (const field of fields) {
    const sectionBytes = count * field.components * TYPE_SIZES[field.type];
    const end = field.byteOffset + sectionBytes;
    if (field.byteOffset < headerSize || end > buffer.byteLength) {
      throw new Error(
        `RUXP: field '${field.name}' section [${field.byteOffset}, ${end}) ` +
          `is outside the payload [${headerSize}, ${buffer.byteLength})`,
      );
    }
    // Checked rather than left to the typed-array constructor, which throws a
    // bare `RangeError` naming neither the field nor the format. The layout
    // guarantees alignment (`header_size` is a multiple of 8 and every f32/u32
    // section is 4-byte sized), so tripping this means the page is malformed.
    const alignment = TYPE_SIZES[field.type];
    if (field.byteOffset % alignment !== 0) {
      throw new Error(
        `RUXP: field '${field.name}' byte_offset ${field.byteOffset} is not ` +
          `${alignment}-byte aligned`,
      );
    }
  }

  const views = new Map<string, RuxpView>();
  return {
    version,
    count,
    offset,
    total,
    fields,
    view(name: string): RuxpView | null {
      const cached = views.get(name);
      if (cached) return cached;
      const field = fields.find((entry) => entry.name === name);
      if (!field) return null;
      const view = makeView(buffer, field, count);
      views.set(name, view);
      return view;
    },
  };
}

/** The field named `name`, or `null`. Lets a caller check a type before viewing. */
export function ruxpField(page: RuxpPage, name: string): RuxpField | null {
  return page.fields.find((field) => field.name === name) ?? null;
}

/**
 * The zero-copy step, and the only reason this format is planar.
 *
 * Every bound and every alignment constraint was checked before this ran, so
 * the constructor cannot throw here.
 */
function makeView(buffer: ArrayBuffer, field: RuxpField, count: number): RuxpView {
  const length = count * field.components;
  switch (field.type) {
    case 'f32':
      return new Float32Array(buffer, field.byteOffset, length);
    case 'u8':
      return new Uint8Array(buffer, field.byteOffset, length);
    case 'u32':
      return new Uint32Array(buffer, field.byteOffset, length);
  }
}

function typeFromCode(code: number, fieldName: string): RuxpFieldType {
  switch (code) {
    case 1:
      return 'f32';
    case 2:
      return 'u8';
    case 3:
      return 'u32';
    default:
      throw new Error(`RUXP: field '${fieldName}' has unknown type code ${code}`);
  }
}

/** ASCII name out of an 8-byte, NUL-padded (never NUL-terminated at 8) slot. */
function readName(buffer: ArrayBuffer, byteOffset: number): string {
  const bytes = new Uint8Array(buffer, byteOffset, 8);
  let end = 0;
  while (end < bytes.length && bytes[end] !== 0) end += 1;
  return String.fromCharCode(...bytes.subarray(0, end));
}

/**
 * A `u64` header field, as a `number`.
 *
 * Read as two `u32`s rather than with `getBigUint64`. Every consumer of
 * `offset`/`total` does `number` arithmetic — page planning, a progress
 * fraction — so a `BigInt` would be converted away immediately, and the pair of
 * `u32`s makes the range guard exact instead of approximate: a value that does
 * not fit a double's 53-bit mantissa is precisely one whose high word exceeds
 * 2^21 - 1.
 */
function readU64(header: DataView, byteOffset: number, what: string): number {
  const low = header.getUint32(byteOffset, true);
  const high = header.getUint32(byteOffset + 4, true);
  if (high > 0x1f_ffff) {
    throw new Error(`RUXP: ${what} exceeds Number.MAX_SAFE_INTEGER (high word 0x${high.toString(16)})`);
  }
  return high * 2 ** 32 + low;
}

/** First bytes of a non-RUXP buffer, for an error message worth reading. */
function describeStart(buffer: ArrayBuffer): string {
  if (buffer.byteLength === 0) return 'empty';
  const bytes = new Uint8Array(buffer, 0, Math.min(4, buffer.byteLength));
  const hex = Array.from(bytes, (byte) => byte.toString(16).padStart(2, '0')).join(' ');
  return `[${hex}]`;
}

File: apps/rux/src/gui/binary_points.cpp
// SPDX-FileCopyrightText: 2026 Povl Filip Sonne-Frederiksen
//
// SPDX-License-Identifier: GPL-3.0-or-later

#include "gui/binary_points.hpp"

#include <algorithm>
#include <array>
#include <cstring>
#include <limits>
#include <stdexcept>
#include <string>

namespace rux::gui {
namespace {

/// One RUXP field, plus how to lift it out of a stored record.
struct FieldSpec {
  const char *name;
  RuxpType type;
  uint8_t components;
  /// Byte offset of this field's first byte within a stored record.
  uint32_t record_offset;
  /// True when the field's bytes must be reordered per point rather than
  /// copied straight out of the record (see the RGB swizzle below).
  bool swizzle_bgr = false;
};

size_t size_of(RuxpType type) {
  switch (type) {
  case RuxpType::f32:
  case RuxpType::u32:
    return 4;
  case RuxpType::u8:
    return 1;
  }
  return 0;
}

/// The field table from docs/gui/binary-points.md, keyed by cloud type.
///
/// Storage record layouts are the ones ProjectDB writes (see the serialize*
/// helpers in libs/reusex/src/core/ProjectDB.cpp):
///   PointXYZRGB — x,y,z as f32 then the packed `rgba` word   (16 B)
///   PointXYZ    — x,y,z as f32                               (12 B)
///   Normal      — nx,ny,nz as f32 then curvature             (16 B)
///   Label       — one u32                                    ( 4 B)
std::vector<FieldSpec> fields_for(std::string_view point_type) {
  if (point_type == "PointXYZRGB")
    return {{"xyz", RuxpType::f32, 3, 0},
            // The stored word at byte 12 is pcl::PointXYZRGB::rgba, and PCL
            // declares that union member over `struct { uint8_t b, g, r, a; }`
            // (PCL_ADD_UNION_RGB in pcl/impl/point_types.hpp). ProjectDB
            // memcpy's the uint32 verbatim, so on a little-endian host the
            // stored bytes are b,g,r,a in that order: b at record offset 12,
            // g at 13, r at 14. RUXP emits r,g,b, so the section starts at 14
            // and walks backwards — hence the swizzle rather than a copy.
            {"rgb", RuxpType::u8, 3, 12, /*swizzle_bgr=*/true}};
  if (point_type == "PointXYZ")
    return {{"xyz", RuxpType::f32, 3, 0}};
  if (point_type == "Normal")
    // Only the first three floats: curvature is stored but deliberately not
    // exposed by either wire format.
    return {{"normal", RuxpType::f32, 3, 0}};
  if (point_type == "Label")
    return {{"label", RuxpType::u32, 1, 0}};
  return {};
}

// Explicit little-endian writes. The header must be byte-identical on any
// host, so nothing here may depend on the host's byte order.

void put_u8(std::vector<uint8_t> &out, size_t at, uint8_t value) {
  out[at] = value;
}

void put_u16(std::vector<uint8_t> &out, size_t at, uint16_t value) {
  out[at] = static_cast<uint8_t>(value & 0xFF);
  out[at + 1] = static_cast<uint8_t>((value >> 8) & 0xFF);
}

void put_u32(std::vector<uint8_t> &out, size_t at, uint32_t value) {
  for (size_t i = 0; i < 4; ++i)
    out[at + i] = static_cast<uint8_t>((value >> (8 * i)) & 0xFF);
}

void put_u64(std::vector<uint8_t> &out, size_t at, uint64_t value) {
  for (size_t i = 0; i < 8; ++i)
    out[at + i] = static_cast<uint8_t>((value >> (8 * i)) & 0xFF);
}

} // namespace

bool ruxp_supports(std::string_view point_type) {
  return !fields_for(point_type).empty();
}

std::vector<uint8_t> encode_ruxp(const reusex::ProjectDB::CloudPage &page) {
  const auto fields = fields_for(page.point_type);
  if (fields.empty())
    throw std::runtime_error("RUXP has no field layout for point type '" +
                             page.point_type + "'");

  const size_t count = static_cast<size_t>(page.count);
  const size_t step = page.point_step;
  if (page.data.size() < count * step)
    throw std::runtime_error(
        "RUXP page is short: " + std::to_string(page.data.size()) +
        " bytes for " + std::to_string(count) + " records of " +
        std::to_string(step));

  const size_t header_size = kRuxpHeaderBase + kRuxpFieldSize * fields.size();
  // header_size is a u16 on the wire, so field_count can never exceed 4093.
  // Nothing today comes near it (the largest table is two fields), but a
  // silent truncation here would produce a page that parses and is wrong.
  if (header_size > std::numeric_limits<uint16_t>::max())
    throw std::runtime_error("RUXP header does not fit a u16: " +
                             std::to_string(fields.size()) + " fields");

  size_t body = header_size;
  std::vector<size_t> section_offset;
  section_offset.reserve(fields.size());
  for (const auto &field : fields) {
    section_offset.push_back(body);
    body += count * field.components * size_of(field.type);
  }

  std::vector<uint8_t> out(body, 0);

  // ---- header ----
  static constexpr std::array<uint8_t, 4> kMagic = {'R', 'U', 'X', 'P'};
  std::memcpy(out.data(), kMagic.data(), kMagic.size());
  put_u16(out, 4, kRuxpVersion);
  put_u16(out, 6, static_cast<uint16_t>(header_size));
  put_u32(out, 8, 0); // flags
  put_u32(out, 12, static_cast<uint32_t>(fields.size()));
  put_u32(out, 16, static_cast<uint32_t>(count));
  put_u32(out, 20, 0); // reserved
  put_u64(out, 24, page.offset);
  put_u64(out, 32, page.total);

  // ---- field descriptors ----
  for (size_t f = 0; f < fields.size(); ++f) {
    const auto &field = fields[f];
    const size_t at = kRuxpHeaderBase + kRuxpFieldSize * f;
    // char[8], NUL-padded; never NUL-terminated when the name is 8 chars.
    const std::string_view name(field.name);
    std::memcpy(out.data() + at, name.data(), std::min<size_t>(name.size(), 8));
    put_u8(out, at + 8, static_cast<uint8_t>(field.type));
    put_u8(out, at + 9, field.components);
    put_u16(out, at + 10, 0); // reserved
    // Absolute from the start of the buffer, so a client needs no arithmetic.
    put_u32(out, at + 12, static_cast<uint32_t>(section_offset[f]));
  }

  // ---- planar payload ----
  for (size_t f = 0; f < fields.size(); ++f) {
    const auto &field = fields[f];
    const size_t width = field.components * size_of(field.type);
    uint8_t *dst = out.data() + section_offset[f];
    const uint8_t *src = page.data.data() + field.record_offset;

    if (field.swizzle_bgr) {
      for (size_t i = 0; i < count; ++i, dst += width, src += step) {
        dst[0] = src[2]; // r
        dst[1] = src[1]; // g
        dst[2] = src[0]; // b
      }
    } else {
      for (size_t i = 0; i < count; ++i, dst += width, src += step)
        std::memcpy(dst, src, width);
    }
  }

  return out;
}

} // namespace rux::gui
